pkg/cvo/status_history_test.go

```diff
@@ -0,0 +1,819 @@
+package cvo
+
+import (
+	"reflect"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/klog/v2"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+func Test_prune(t *testing.T) {
+	tests := []struct {
+		name       string
+		history    []configv1.UpdateHistory
+		maxHistory int
+		want       []configv1.UpdateHistory
+	}{
+		{
+			name: "partial update within a minor transition",
+			history: []configv1.UpdateHistory{
+				{State: configv1.CompletedUpdate, Version: "4.6.3"},
+				{State: configv1.CompletedUpdate, Version: "4.5.3"},
+				{State: configv1.CompletedUpdate, Version: "4.4.3"},
+				{State: configv1.CompletedUpdate, Version: "4.3.3"},
+				{State: configv1.CompletedUpdate, Version: "4.3.2"},
+				{State: configv1.CompletedUpdate, Version: "4.3.1"},
+				{State: configv1.CompletedUpdate, Version: "4.2.1"},
+				{State: configv1.PartialUpdate, Version: "4.1.4"},
+				{State: configv1.PartialUpdate, Version: "4.1.3"},
+				{State: configv1.PartialUpdate, Version: "4.1.2"},
+				{State: configv1.CompletedUpdate, Version: "4.1.1"},
+			},
+			maxHistory: 10,
+			want: []configv1.UpdateHistory{
+				{State: configv1.CompletedUpdate, Version: "4.6.3"},
+				{State: configv1.CompletedUpdate, Version: "4.5.3"},
+				{State: configv1.CompletedUpdate, Version: "4.4.3"},
+				{State: configv1.CompletedUpdate, Version: "4.3.3"},
+				{State: configv1.CompletedUpdate, Version: "4.3.2"},
+				{State: configv1.CompletedUpdate, Version: "4.3.1"},
+				{State: configv1.CompletedUpdate, Version: "4.2.1"},
+				{State: configv1.PartialUpdate, Version: "4.1.4"},
+				{State: configv1.PartialUpdate, Version: "4.1.3"},
+				{State: configv1.CompletedUpdate, Version: "4.1.1"},
+			},
+		},
+		{
+			name: "partial update within a z-stream transition",
+			history: []configv1.UpdateHistory{
+				{State: configv1.CompletedUpdate, Version: "4.5.3"},
+				{State: configv1.CompletedUpdate, Version: "4.4.3"},
+				{State: configv1.CompletedUpdate, Version: "4.4.3"},
+				{State: configv1.CompletedUpdate, Version: "4.4.2"},
+				{State: configv1.CompletedUpdate, Version: "4.1.10"},
+				{State: configv1.CompletedUpdate, Version: "4.1.9"},
+				{State: configv1.CompletedUpdate, Version: "4.1.3"},
```

Comment:
nit: having a partial update followed by a completed update with the same target version is unlikely.    Can we use 4.1.4 for this line?

Reply:
It now doesn't matter to the tested code. It did when I was explicitly checking for a "same z-stream" version transition. But I thought this would be fairly common - i.e. my upgrade to 4.1.3 didn't complete (is Partial), oh, let me try again?

Reply:
if you are on A, attempt an update to B, give up (returning to A), and then attempt B again, you get ABAB history, not ABB history.  But :shrug: I don't have strong feelings about what goes into your unit test-case here.

Reply:
I did change it as you suggested so no problem there. Yeah, I see what you're saying now about the ABAB history.